fred2/campaignfilelistbox.cpp

```diff
@@ -46,38 +46,57 @@ END_MESSAGE_MAP()
 /////////////////////////////////////////////////////////////////////////////
 // campaign_filelist_box message handlers
 
-void campaign_filelist_box::initialize()
+void campaign_filelist_box::initialize(const CString &path)
 {
-	int i, z, num_files;
-	char *mission_filenames[2000];
-	char mission_filenames_arr[2000][MAX_FILENAME_LEN];
+	int i, z;
+	SCP_vector<SCP_string> mission_filenames;
 	mission a_mission;
 
 	ResetContent();
 
 	extern int Skip_packfile_search;
 	Skip_packfile_search = 1;
-	num_files = cf_get_file_list_preallocated(2000, mission_filenames_arr, mission_filenames, CF_TYPE_MISSIONS, "*.fs2");
+
+	if (path.IsEmpty())
+	{
+		cf_get_file_list(mission_filenames, CF_TYPE_MISSIONS, "*.fs2");
+
+		for (auto &filename: mission_filenames)
+			filename += ".fs2";
+	}
+	else
+	{
+		SCP_vector<file_list_full_info> files;
+		cf_get_list_of_files((LPCTSTR)path, files, "*.fs2");
+
+		for (const auto &file: files)
+			mission_filenames.push_back(file.name);
+	}
+
```

Comment:
The changes to `cf_get_file_list` I made mean you can just do something like:

```
SCP_string filter;

if (!path.IsEmpty()) {
    filter = (LPCTSTR)path;
    filter += DIR_SEPARATOR_CHAR;
}

filter += "*.fs2";

cf_get_file_list(mission_filenames, CF_TYPE_MISSIONS, filter.c_str());
```

and it should work in either case with less code. Basically if the `filter` argument is an absolute path then it will split off the actual filter and only return matching files in that exact path. Then you would haven't to make that internal `list_of_files` global.

Reply:
Gotcha